Implement GitOpsDeploymentRepositoryCredential API CR in GitOps Service

drpaneas opened the pull request:
#### Description:
- Processing logic for Backend controller that is responsible for acting upon a  GitOpsDeploymentRepositoryCredential CR.

#### Link to JIRA Story (if applicable):

https://issues.redhat.com/browse/GITOPSRVCE-96


Instruction for the reviewer:

- There were a number of problems with logging. That is why you will find many replacements of `log` and `logger` into `l`.
- The main that has he logic is the `sharedresourceloop.go`
- The test file is the `sharedresourceloop_test.go` the last `It` context.
- What is missing is that the function `internalProcessMessage_ReconcileRepositoryCredential` is not getting triggered by the backend controller when you create a real `GitOpsDeploymentRepositoryCredential` CR. Most likely this is related to the code relocation from the file [application_event_runner_deployments.go](https://github.com/redhat-appstudio/managed-gitops/pull/239/files#diff-c77031be4bccf13100733b7e5389de32b45fb0ab49c8f586942cff535b3f4784)

To run the tests with the GinkgoWritter logger: `go test -v  ./eventloop/shared_resource_loop -args -ginkgo.v "GitOpsDeploymentRepositoryCredential"`

logs:
```
  Begin Captured GinkgoWriter Output >>
    1.6656705627196548e+09	DEBUG	sharedResourceEventLoop received message: getGitopsEngineInstanceById	{"workspace": "923beaa6-a156-4db5-b7f6-533288a101c5"}
    1.665670562721324e+09	DEBUG	sharedResourceEventLoop received message: getOrCreateClusterUserByNamespaceUID	{"workspace": "923beaa6-a156-4db5-b7f6-533288a101c5"}
    1.665670562724163e+09	INFO	Created Cluster User with User ID: c240706e-1508-4b2f-b124-847737f60ea3	{"clusteruser_id": "c240706e-1508-4b2f-b124-847737f60ea3", "user_name": "923beaa6-a156-4db5-b7f6-533288a101c5"}
    1.6656705627242088e+09	DEBUG	sharedResourceEventLoop received message: getOrCreateClusterUserByNamespaceUID	{"workspace": "923beaa6-a156-4db5-b7f6-533288a101c5"}
    1.6656705627373629e+09	INFO	Created Cluster Credentials for GitOpsEngineCluster: 61ccec9f-baf8-4f14-9e59-6be42c0c6f17	{"host": "host", "kube-config-length": 11, "kube-config-context": 19, "serviceaccount_ns": "serviceaccount_ns", "serviceaccount-bearer-token-length": 27}
    1.665670562739448e+09	INFO	Created GitopsEngineCluster: 8bcc64a4-b449-4c95-80c4-7e47b82b4beb	{"clustercredentials_id": "61ccec9f-baf8-4f14-9e59-6be42c0c6f17", "gitopsenginecluster_id": "8bcc64a4-b449-4c95-80c4-7e47b82b4beb"}
    1.6656705627421172e+09	INFO	Created KubernetesResourceToDBResourceMapping with DBRelationKey: 8bcc64a4-b449-4c95-80c4-7e47b82b4beb	{"k8sResourceType": "Namespace", "k8sResourceUID": "df5c3c96-532c-47e2-97ca-95df0a3c393d", "dbRelationType": "GitopsEngineCluster", "dbRelationKey": "8bcc64a4-b449-4c95-80c4-7e47b82b4beb"}
    1.665670562744859e+09	INFO	Created GitopsEngineInstance: ff3ab762-b007-4424-9e12-b2fa69c20a60
    1.6656705627466161e+09	INFO	Created KubernetesResourceToDBResourceMapping with KubernetesResourceUID: df5c3c96-532c-47e2-97ca-95df0a3c393d	{"k8sResourceType": "Namespace", "k8sResourceUID": "df5c3c96-532c-47e2-97ca-95df0a3c393d", "dbRelationType": "GitopsEngineInstance", "dbRelationKey": "ff3ab762-b007-4424-9e12-b2fa69c20a60"}
    1.665670562746634e+09	INFO	Attempting to get the CR
    1.6656705627537699e+09	INFO	Attempting to get the CR
    1.665670562758892e+09	INFO	Creating operation...
    1.665670562764606e+09	INFO	Created Operation database row
    1.665670562764638e+09	INFO	Created K8s Operation CR
    1.665670562764643e+09	INFO	operationCR	{"operationCR": {"metadata":{"name":"operation-f0943aa8-2b57-4b02-9e56-31729a8f5946","namespace":"my-user","resourceVersion":"1","creationTimestamp":null},"spec":{"operationID":"f0943aa8-2b57-4b02-9e56-31729a8f5946"},"status":{}}}
    1.66567056276468e+09	INFO	operationDB	{"operationDB": {"Operation_id":"f0943aa8-2b57-4b02-9e56-31729a8f5946","Instance_id":"ff3ab762-b007-4424-9e12-b2fa69c20a60","Resource_id":"test-gitopsdeploymenrepositorycredential","Operation_owner_user_id":"c240706e-1508-4b2f-b124-847737f60ea3","Resource_type":"RepositoryCredentials","Created_on":"2022-10-13T16:16:02.762078+02:00","Last_state_update":"2022-10-13T16:16:02.762078+02:00","State":"Waiting","Human_readable_state":"","SeqID":47,"GC_expiration_time":0}}
    1.665670562773057e+09	INFO	Attempting to get the CR
    1.6656705627832398e+09	INFO	Attempting to get the CR
    1.6656705627845278e+09	INFO	Secret name changed	{"old": "test-secret-2", "new": "test-secret"}
    1.665670562784543e+09	INFO	Syncing with database...
    1.665670562786245e+09	INFO	Creating operation...
    1.665670562787413e+09	INFO	Skipping Operation creation, as it already exists for resource.	{"existingOperationState": "Waiting"}
    1.665670562787424e+09	INFO	operationCR	{"operationCR": {"kind":"Operation","apiVersion":"managed-gitops.redhat.com/v1alpha1","metadata":{"name":"operation-f0943aa8-2b57-4b02-9e56-31729a8f5946","namespace":"my-user","resourceVersion":"1","creationTimestamp":null},"spec":{"operationID":"f0943aa8-2b57-4b02-9e56-31729a8f5946"},"status":{}}}
    1.66567056278744e+09	INFO	operationDB	{"operationDB": {"Operation_id":"f0943aa8-2b57-4b02-9e56-31729a8f5946","Instance_id":"ff3ab762-b007-4424-9e12-b2fa69c20a60","Resource_id":"test-gitopsdeploymenrepositorycredential","Operation_owner_user_id":"c240706e-1508-4b2f-b124-847737f60ea3","Resource_type":"RepositoryCredentials","Created_on":"2022-10-13T14:16:02.762078Z","Last_state_update":"2022-10-13T14:16:02.762078Z","State":"Waiting","Human_readable_state":"","SeqID":47,"GC_expiration_time":0}}
    1.6656705627955399e+09	INFO	Attempting to get the CR
    1.665670562795558e+09	INFO	Error getting the CR
    1.665670562796917e+09	INFO	[Leftover] RepositoryCredential row found in DB
    1.665670562798645e+09	DEBUG	Deleted Repository Credential from the database	{"repocred id": "test-gitopsdeploymenrepositorycredential"}
    1.665670562798662e+09	INFO	RepositoryCredential row deleted from DB
    1.665670562798665e+09	INFO	Creating an Operation for the deleted RepositoryCredential DB row
    1.665670562798667e+09	INFO	Creating operation...
    1.665670562802567e+09	INFO	Created Operation database row
    1.66567056280259e+09	INFO	Created K8s Operation CR
    1.665670562802593e+09	INFO	operationCR	{"operationCR": {"metadata":{"name":"operation-7fa6d9f1-e31e-4df0-afcb-a30c3696276a","namespace":"my-user","resourceVersion":"1","creationTimestamp":null},"spec":{"operationID":"7fa6d9f1-e31e-4df0-afcb-a30c3696276a"},"status":{}}}
    1.66567056280261e+09	INFO	operationDB	{"operationDB": {"Operation_id":"7fa6d9f1-e31e-4df0-afcb-a30c3696276a","Instance_id":"ff3ab762-b007-4424-9e12-b2fa69c20a60","Resource_id":"test-gitopsdeploymenrepositorycredential","Operation_owner_user_id":"c240706e-1508-4b2f-b124-847737f60ea3","Resource_type":"RepositoryCredentials","Created_on":"2022-10-13T16:16:02.800886+02:00","Last_state_update":"2022-10-13T16:16:02.800886+02:00","State":"Waiting","Human_readable_state":"","SeqID":48,"GC_expiration_time":0}}
    1.665670562802621e+09	INFO	Operation created for the deleted RepositoryCredential DB row
    1.665670562810223e+09	INFO	Attempting to get the CR
    1.665670562810238e+09	INFO	Error getting the CR
    1.665670562811154e+09	INFO	DB entry not found
    1.665670562811162e+09	INFO	No leftovers. Nothing to do
```

drpaneas commented:
/retest

drpaneas commented:
/retest-required

drpaneas commented:
/retest-required

jgwest reviewed (changes requested):
Hey @drpaneas , one thing I noticed is that the RepositoryCredentials primary key should be a randomly generated UID, not the uid/name of the resource repository credential CR.

We then use this UID in the APICRToDatabaseMapping table, to track the relationship from the K8s resource in KCP/K8s <-> database resource (repository credentials row). This is described here: https://docs.google.com/document/d/1e1UwCbwK-Ew5ODWedqp_jZmhiZzYWaxEvIL-tqebMzo/edit#heading=h.45brv1rx6wmo
(Let me know if you have any questions, or if the documentation doesn't do a good job of explaining the concept).

This allows us flexibility in tracking the relationship between CRs <-> database rows, across multiple namespaces/KCP workspaces, and the ability to add additional fields to the relationship table (APICRToDBMapping), without needing to add them to the RepositoryCredential table.

Pseudocode for how this all fits together can be found here:
- https://github.com/redhat-appstudio/managed-gitops/compare/main...jgwest:managed-gitops:shared-resource-loop-repo-cred-sept-2022?expand=1
- under 'internalProcessMessage_ReconcileRepositoryCredential' in sharedresourceloop.go

You can see examples of how to use APICRToDatabaseMapping for GitOpsDeploymentSyncRun in application_event_runner_syncruns, in 'applicationEventRunner_handleSyncRunModifiedInternal'.
Or for GitOpsDeploymentManagedEnvironment in sharedresourceloop_managedenv.go, in 'internalProcessMessage_ReconcileSharedManagedEnv'.

You may ask: in the `internalProcessMessage_ReconcileRepositoryCredential(...)` function, how do I get the APICRToDatabaseMapping if all I have is the CR name/namespace/namespace UID?

```go
apiCRToDBMapping := db.APICRToDatabaseMapping{
	APIResourceType: db.APICRToDatabaseMapping_ResourceType_RepositoryCredential,
	APIResourceUID:  string(repositoryCredentialCR.UID),
	DBRelationType:  db.APICRToDatabaseMapping_DBRelationType_RepositoryCredential,
}

if err := dbQueries.GetDatabaseMappingForAPICR(ctx, &apiCRToDBMapping); err != nil { /* ... */ }
```

Here's [an example from SyncRun](https://github.com/jgwest/managed-gitops/blob/9164b7b435b0687e05a219fccceb983ca2c1a872/backend/eventloop/application_event_loop/application_event_runner_syncruns.go#L57). (Note that GitOpsDeploymentSyncRun CR points to a SyncOperation DB row, just as GitOpsDeploymentRepositoryCredentials points to a RepositoryCredentials DB row), and from [ManagedEnvironment](https://github.com/redhat-appstudio/managed-gitops/blob/ff3fa7a343e989b7a3a9fcf19c0736fe5f088d90/backend/eventloop/shared_resource_loop/sharedresourceloop_managedenv.go#L72).


You will likewise need to ensure that the APICRToDatabaseMapping is created when the RepositoryCredential row is created. Here's [an example from SyncRun](https://github.com/jgwest/managed-gitops/blob/9164b7b435b0687e05a219fccceb983ca2c1a872/backend/eventloop/application_event_loop/application_event_runner_syncruns.go#L209), or from [ManagedEnvironment](https://github.com/redhat-appstudio/managed-gitops/blob/ff3fa7a343e989b7a3a9fcf19c0736fe5f088d90/backend/eventloop/shared_resource_loop/sharedresourceloop_managedenv.go#L472).

Finally, you will need to delete old RepositoryCredential rows and APICRToDatabaseMappings that point to them. You can see [an example from SyncRun](https://github.com/jgwest/managed-gitops/blob/9164b7b435b0687e05a219fccceb983ca2c1a872/backend/eventloop/application_event_loop/application_event_runner_syncruns.go#L333), or [from ManagedEnvironment](https://github.com/redhat-appstudio/managed-gitops/blob/ff3fa7a343e989b7a3a9fcf19c0736fe5f088d90/backend/eventloop/shared_resource_loop/sharedresourceloop_managedenv.go#L247).

Once this is done I will review the unit tests, and remaining functions.

drpaneas commented:
Hey @jgwest  thank you very much for the PR review. I've addressed all your comments (apart from the verbose log.Info() statements -- I will do that, as soon as the tests are passing, since I am using them for debugging atm). Something doesn't seem to be working here.

Here's my logs:

```
------------------------------
• [FAILED] [0.131 seconds]
SharedResourceEventLoop Test
/Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/backend/eventloop/shared_resource_loop/sharedresourceloop_test.go:32
  Shared Resource Event Loop test
  /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/backend/eventloop/shared_resource_loop/sharedresourceloop_test.go:43
    [It] Should fetch a GitOpsDeploymentRepositoryCredential.
    /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/backend/eventloop/shared_resource_loop/sharedresourceloop_test.go:273

  Begin Captured GinkgoWriter Output >>
    1.667236596399415e+09	DEBUG	sharedResourceEventLoop received message: getGitopsEngineInstanceById	{"workspace": "1a148a3c-d5c2-4321-a00a-2b77ee9b841d"}
    1.6672365964008121e+09	DEBUG	sharedResourceEventLoop received message: getOrCreateClusterUserByNamespaceUID	{"workspace": "1a148a3c-d5c2-4321-a00a-2b77ee9b841d"}
    1.667236596403955e+09	INFO	Created Cluster User with User ID: 8d48abfe-8f8d-4389-a6c3-d9a8a03abb75	{"clusteruser_id": "8d48abfe-8f8d-4389-a6c3-d9a8a03abb75", "user_name": "1a148a3c-d5c2-4321-a00a-2b77ee9b841d"}
    1.6672365964039822e+09	DEBUG	sharedResourceEventLoop received message: getOrCreateClusterUserByNamespaceUID	{"workspace": "1a148a3c-d5c2-4321-a00a-2b77ee9b841d"}
    1.6672365964189599e+09	INFO	Created Cluster Credentials for ManagedEnvironment: 2f3daf66-27e3-4fd6-a35f-bafa55d5dfec	{"host": "host", "kube-config-length": 11, "kube-config-context": 19, "serviceaccount_ns": "serviceaccount_ns", "serviceaccount-bearer-token-length": 27}
    1.667236596423538e+09	INFO	Created Managed Environment: b103dba6-6c67-45f8-b20f-777b081e8e13	{"clusterCredentialsID": "2f3daf66-27e3-4fd6-a35f-bafa55d5dfec", "managedEnvironmentID": "b103dba6-6c67-45f8-b20f-777b081e8e13", "managedEnvironmentName": "Managed Environment for my-user"}
    1.667236596426639e+09	INFO	Created KubernetesResourceToDBResourceMapping for ManagedEnvironment: 1a148a3c-d5c2-4321-a00a-2b77ee9b841d	{"k8sResourceType": "Namespace", "k8sResourceUID": "1a148a3c-d5c2-4321-a00a-2b77ee9b841d", "dbRelationType": "ManagedEnvironment", "dbRelationKey": "b103dba6-6c67-45f8-b20f-777b081e8e13"}
    1.667236596429804e+09	INFO	Created Cluster Credentials for GitOpsEngineCluster: c3cbc9f8-83d5-4bad-baf2-c6f809af59f6	{"host": "host", "kube-config-length": 11, "kube-config-context": 19, "serviceaccount_ns": "serviceaccount_ns", "serviceaccount-bearer-token-length": 27}
    1.6672365964315221e+09	INFO	Created GitopsEngineCluster: 3753555f-feb4-4ec5-8e99-f548297342c9	{"clustercredentials_id": "c3cbc9f8-83d5-4bad-baf2-c6f809af59f6", "gitopsenginecluster_id": "3753555f-feb4-4ec5-8e99-f548297342c9"}
    1.667236596433172e+09	INFO	Created KubernetesResourceToDBResourceMapping with DBRelationKey: 3753555f-feb4-4ec5-8e99-f548297342c9	{"k8sResourceType": "Namespace", "k8sResourceUID": "5357e60b-46e8-44cf-8345-d9f233185b18", "dbRelationType": "GitopsEngineCluster", "dbRelationKey": "3753555f-feb4-4ec5-8e99-f548297342c9"}
    1.667236596436841e+09	INFO	Created GitopsEngineInstance: ff3169b1-7796-4e73-91b0-3b760f961d71
    1.667236596438953e+09	INFO	Created KubernetesResourceToDBResourceMapping with KubernetesResourceUID: 5357e60b-46e8-44cf-8345-d9f233185b18	{"k8sResourceType": "Namespace", "k8sResourceUID": "5357e60b-46e8-44cf-8345-d9f233185b18", "dbRelationType": "GitopsEngineInstance", "dbRelationKey": "ff3169b1-7796-4e73-91b0-3b760f961d71"}
    1.6672365964389691e+09	INFO	Attempting to get the CR
    1.667236596452949e+09	INFO	Attempting to get the CR
    1.6672365964614668e+09	INFO	APICRToDatabaseMapping list	{"list": null}
    1.667236596461523e+09	INFO	Before the loop
    1.667236596461525e+09	INFO	After the loop
  << End Captured GinkgoWriter Output

  Expected
      <*errors.errorString | 0xc00033dde0>: {
          s: "no repository credentials found in the database for the CR",
      }
  to be nil
  In [It] at: /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/backend/eventloop/shared_resource_loop/sharedresourceloop_test.go:393
```

drpaneas commented:
If it helps, I've managed to fish out the error:

```
err	{"error": "unable to retrieve APICRToDatabase mapping for GitOpsDeploymentRepositoryCredential:19040c2d-56c7-4dee-ad88-a4d8d6926445: no rows in result set"}
```

drpaneas commented:
Ah, I fixed the problem. The list was nil so it wasn't getting in the loop. Now I have a different error:

```  
Expected
      <*errors.errorString | 0xc00109ea30>: {s: "invalid pk"}
  to be nil
  In [It] at: /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/backend/eventloop/shared_resource_loop/sharedresourceloop_test.go:409
```


which I think is because my Unit Test is wrong:

```go
var operationDB db.Operation
			operationDB.Resource_id = dbRepoCred.RepositoryCredentialsID
```

I think this assumption is wrong. Right?

jgwest commented:
@drpaneas your assumption is correct, the resource ID field points to the primary key of the RepositoryCredentials row in the database (or the primary key of whatever resource table is being referenced: the Operation DB table is 'generic')

So in this section of code, it looks like you are looking for an Operation that points to a particular RepositoryCredentials. You can use `ListOperationsByResourceIdAndTypeAndOwnerId` for this, e.g.
```go
dbQueries.ListOperationsByResourceIdAndTypeAndOwnerId(ctx, /* primary key of repository credential table*/, db.OperationResourceType_RepositoryCredential,
		&operations, /* cluster user, the test likely has this set in a local variable */); err != nil { /* .. */ }
```

drpaneas commented:
Ok, I did that and it worked. New error now:


```
"unable to create repository credential in the database: cannot create repository credentials: ERROR #23505 duplicate key value violates unique constraint \"repositorycredentials_pkey\"",
```

It tries to create the DB Row twice and it obviously fails (for a good reason). The problem is that the second time it should find the DB Row. Apparently it doesn't, so it tries to create it. It already exists though, thus it fails. See the logs:

```
    1.667250932422673e+09	INFO	Created KubernetesResourceToDBResourceMapping with KubernetesResourceUID: 163ec500-52a7-47e3-b6ec-4d61de17148f	{"k8sResourceType": "Namespace", "k8sResourceUID": "163ec500-52a7-47e3-b6ec-4d61de17148f", "dbRelationType": "GitopsEngineInstance", "dbRelationKey": "5be01b6b-ae49-4b69-b80f-7ec1c844c9af"}
    1.667250932422687e+09	INFO	Attempting to get the CR
    1.667250932422763e+09	INFO	CR found	{"CR Name": "test-gitopsdeploymenrepositorycredential", "Namespace": "my-user"}
    1.667250932422768e+09	INFO	gitopsDeploymentRepositoryCredentialCR	{"gitopsDeploymentRepositoryCredentialCR": {"kind":"GitOpsDeploymentRepositoryCredential","apiVersion":"managed-gitops.redhat.com/v1alpha1","metadata":{"name":"test-gitopsdeploymenrepositorycredential","namespace":"my-user","uid":"adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95","resourceVersion":"1","creationTimestamp":null},"spec":{"repository":"https://fakegithub.com/test/test-repository","secret":"test-secret"},"status":{}}}
    1.6672509324227872e+09	INFO	UID	{"UID": "adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95"}
    1.667250932434827e+09	INFO	Attempting to get the CR
    1.667250932434905e+09	INFO	CR found	{"CR Name": "test-gitopsdeploymenrepositorycredential", "Namespace": "my-user"}
    1.667250932434911e+09	INFO	gitopsDeploymentRepositoryCredentialCR	{"gitopsDeploymentRepositoryCredentialCR": {"kind":"GitOpsDeploymentRepositoryCredential","apiVersion":"managed-gitops.redhat.com/v1alpha1","metadata":{"name":"test-gitopsdeploymenrepositorycredential","namespace":"my-user","uid":"adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95","resourceVersion":"1","creationTimestamp":null},"spec":{"repository":"https://fakegithub.com/test/test-repository","secret":"test-secret"},"status":{}}}
    1.6672509324349189e+09	INFO	UID	{"UID": "adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95"}
    1.66725093243504e+09	INFO	Getting mapping for GitOpsDeploymentRepositoryCredential
    1.6672509324402618e+09	INFO	Error getting mapping for GitOpsDeploymentRepositoryCredential
    1.667250932440268e+09	INFO	err	{"error": "unable to retrieve APICRToDatabase mapping for GitOpsDeploymentRepositoryCredential:adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95: no rows in result set"}
    1.667250932440276e+09	INFO	Since DB was not found, we create the DB Row now
    1.6672509324457269e+09	INFO	Created RepositoryCredential in the DB	{"repositoryCredential": {"RepositoryCredentialsID":"test-gitopsdeploymenrepositorycredential","UserID":"74c9e724-0245-4030-a6e4-9491ab538c4b","PrivateURL":"https://fakegithub.com/test/test-repository","AuthUsername":"test-username","AuthPassword":"test-password","AuthSSHKey":"","SecretObj":"test-secret","EngineClusterID":"5be01b6b-ae49-4b69-b80f-7ec1c844c9af","SeqID":1}}
    1.667250932445784e+09	INFO	Creating operation...
    1.667250932454527e+09	INFO	Created Operation database row
    1.667250932454564e+09	INFO	Created K8s Operation CR
    1.667250932454566e+09	INFO	operationCR	{"operationCR": {"metadata":{"name":"operation-f3281eed-c05a-4c3e-91b0-bd9325fa90e2","namespace":"my-user","resourceVersion":"1","creationTimestamp":null},"spec":{"operationID":"f3281eed-c05a-4c3e-91b0-bd9325fa90e2"},"status":{}}}
    1.667250932454619e+09	INFO	operationDB	{"operationDB": {"Operation_id":"f3281eed-c05a-4c3e-91b0-bd9325fa90e2","Instance_id":"5be01b6b-ae49-4b69-b80f-7ec1c844c9af","Resource_id":"test-gitopsdeploymenrepositorycredential","Operation_owner_user_id":"74c9e724-0245-4030-a6e4-9491ab538c4b","Resource_type":"RepositoryCredentials","Created_on":"2022-10-31T22:15:32.449443+01:00","Last_state_update":"2022-10-31T22:15:32.449443+01:00","State":"Waiting","Human_readable_state":"","SeqID":1,"GC_expiration_time":0}}
    1.66725093245739e+09	INFO	Operations	{"operations": [{"Operation_id":"f3281eed-c05a-4c3e-91b0-bd9325fa90e2","Instance_id":"5be01b6b-ae49-4b69-b80f-7ec1c844c9af","Resource_id":"test-gitopsdeploymenrepositorycredential","Operation_owner_user_id":"74c9e724-0245-4030-a6e4-9491ab538c4b","Resource_type":"RepositoryCredentials","Created_on":"2022-10-31T21:15:32.449443Z","Last_state_update":"2022-10-31T21:15:32.449443Z","State":"Waiting","Human_readable_state":"","SeqID":1,"GC_expiration_time":0}]}
    1.667250932457432e+09	INFO	Count	{"count": 1}
    1.667250932457438e+09	INFO	TEST: Get Operation State	{"operation": {"Operation_id":"f3281eed-c05a-4c3e-91b0-bd9325fa90e2","Instance_id":"5be01b6b-ae49-4b69-b80f-7ec1c844c9af","Resource_id":"test-gitopsdeploymenrepositorycredential","Operation_owner_user_id":"74c9e724-0245-4030-a6e4-9491ab538c4b","Resource_type":"RepositoryCredentials","Created_on":"2022-10-31T21:15:32.449443Z","Last_state_update":"2022-10-31T21:15:32.449443Z","State":"Waiting","Human_readable_state":"","SeqID":1,"GC_expiration_time":0}}
    1.6672509324574468e+09	INFO	Get the RepositoryCredential DB row using the operationDB.Resource_id	{"operation Resource ID": "test-gitopsdeploymenrepositorycredential"}
    1.667250932460587e+09	INFO	DB Row Fetched	{"fetch": {"RepositoryCredentialsID":"test-gitopsdeploymenrepositorycredential","UserID":"74c9e724-0245-4030-a6e4-9491ab538c4b","PrivateURL":"https://fakegithub.com/test/test-repository","AuthUsername":"test-username","AuthPassword":"test-password","AuthSSHKey":"","SecretObj":"test-secret","EngineClusterID":"5be01b6b-ae49-4b69-b80f-7ec1c844c9af","SeqID":1}}
    1.667250932460609e+09	INFO	DB Row Original	{"dbRepoCred": {"RepositoryCredentialsID":"test-gitopsdeploymenrepositorycredential","UserID":"74c9e724-0245-4030-a6e4-9491ab538c4b","PrivateURL":"https://fakegithub.com/test/test-repository","AuthUsername":"test-username","AuthPassword":"test-password","AuthSSHKey":"","SecretObj":"test-secret","EngineClusterID":"5be01b6b-ae49-4b69-b80f-7ec1c844c9af","SeqID":1}}
    1.667250932460624e+09	INFO	Re-running the internalProcessMessage_ReconcileRepositoryCredential()
    1.66725093246855e+09	INFO	Attempting to get the CR
    1.6672509324686148e+09	INFO	CR found	{"CR Name": "test-gitopsdeploymenrepositorycredential", "Namespace": "my-user"}
    1.667250932468621e+09	INFO	gitopsDeploymentRepositoryCredentialCR	{"gitopsDeploymentRepositoryCredentialCR": {"kind":"GitOpsDeploymentRepositoryCredential","apiVersion":"managed-gitops.redhat.com/v1alpha1","metadata":{"name":"test-gitopsdeploymenrepositorycredential","namespace":"my-user","uid":"adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95","resourceVersion":"1","creationTimestamp":null},"spec":{"repository":"https://fakegithub.com/test/test-repository","secret":"test-secret"},"status":{}}}
    1.667250932468628e+09	INFO	UID	{"UID": "adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95"}
    1.667250932468662e+09	INFO	Getting mapping for GitOpsDeploymentRepositoryCredential
    1.667250932469569e+09	INFO	Error getting mapping for GitOpsDeploymentRepositoryCredential
    1.667250932469586e+09	INFO	err	{"error": "unable to retrieve APICRToDatabase mapping for GitOpsDeploymentRepositoryCredential:adf8033b-19eb-4cd0-aa1b-67fb7fc1fe95: no rows in result set"}

^^^^^ This is the problem. It should have found it here. ^^^^^^

    1.667250932469595e+09	INFO	Since DB was not found, we create the DB Row now
```


The code that checks that:

```go
var apiCRToDBList []db.APICRToDatabaseMapping
	mapping := db.APICRToDatabaseMapping{
		APIResourceType: db.APICRToDatabaseMapping_ResourceType_GitOpsDeploymentRepositoryCredential,
		// APIResourceUID:       string(repositoryCredentialCRNamespace.GetUID()),
		APIResourceUID: string(gitopsDeploymentRepositoryCredentialCR.UID),
		//APIResourceName:      "",
		//APIResourceNamespace: "",
		//NamespaceUID:         "",
		DBRelationType: db.APICRToDatabaseMapping_DBRelationType_RepositoryCredential,
		// DBRelationKey:        "",
	}

	l.Info("Getting mapping for GitOpsDeploymentRepositoryCredential")
	if err := dbQueries.GetDatabaseMappingForAPICR(ctx, &mapping); err != nil {
```

Do you see anything funky here?

drpaneas commented:
I've fixed the previous error as well. Indeed I had to create the mapping as well (duh) :P 
The unit-tests are now passing fine.

jgwest reviewed (changes requested):
Hey @drpaneas, I've pushed a couple commits to your branch, you'll need to `git pull` to pick up the latest changes. These commits:
- Some miscellaneous tweaks and bug fixing while reviewing
- I've made changes to `internalProcessMessage_ReconcileRepositoryCredential`, to change the logic to clean up APICRToDatabaseMappings first, and then to handle the rest of the logic, which I've found (in other places, such a GitOpsDeployment runner) to produce cleaner code overall.
	
Next steps for this PR:
- See the 'TODO: PANOS' comments I added, for additional changes that are needed.
- Code cleanup: the `internalProcessMessage_ReconcileRepositoryCredential` function can likely be refactored into smaller chunks for clarify (for example, moving the delete, update, and create sections of the code into their own functions)
- It may also be possible to break the unit tests into small chunks, too
- E2E tests
    - Which will make sure the code that you and I wrote actually works, outside of the unit tests :smile:

drpaneas commented on backend/eventloop/shared_resource_loop/sharedresourceloop_test.go:
```diff
@@ -269,5 +269,254 @@ var _ = Describe("SharedResourceEventLoop Test", func() {
+
+			defer dbq.CloseDatabase()
+
+			clusterCredentials := db.ClusterCredentials{
```
Thanks for the input and feedback on this, I will proceed with them ;)
But as for E2E I remember we have discussed it in the past and for this story is difficult to implement (e.g. which private repo to pick, if we have to setup gitlab within the CI, etc). Did something change in the meantime or have an idea how we can put up with a E2E scenario for this?

jgwest replied:
@drpaneas,  re: commits I pushed, here's a link to a branch containing them: https://github.com/redhat-appstudio/managed-gitops/compare/main...jgwest:managed-gitops:96?expand=1

re: E2E test, ah I forgot about that. We had the same issue with applicationset controller in Argo CD. 

For those, here's the best I can think of, that we can do for now:

1) I've created a private git repository in https://github.com/managed-gitops-test-data/private-repo-test
	- It contains a single `ConfigMap` which can be deployed using a GitOpsDeployment
2) We can add our team members to the org, so they can see/access the private repo.
	- I've sent you an invite, as a first step.
3) We can create a new E2E test which attempts to deploy from this private Git repository, using a GitOpsDeploymentRepositoryCredential and a GitOpsDeployment
	- The E2E test should create a GitOpsDeploymentRepositoryCredential using a valid personal access token (PAT), with that PAT being able to read from managed-gitops-test-data/private-repo-test
	- The test should read the personal access token from a new env var 'GITHUB_PERSONAL_ACCESS_TOKEN', that needs to be specified when the E2E test is run (export GITHUB_PERSONAL_ACCESS_TOKEN=(pat for a user with access to the private git repo))
	- The test should create a GitsOpsDeployment that references the GitOpsDeploymentRepositoryCredential
	- The test should pass if the GitOpsDeployment was able to deploy the ConfigMap from the private repo.
	- Finally, at the top of the test, if the 'GITHUB_PERSONAL_ACCESS_TOKEN' variable is not set, we should call `Skip("GITHUB_PERSONAL_ACCESS_TOKEN is not set, so skipping test")`.
	
This at least gets us an E2E test which we can run locally to verify the functionality. (It won't help us run the test as part of PRs, unless/until we can find a way to securely embed the secret in the PR)

drpaneas commented:
@jgwest please new review when you have the time ;)

jgwest commented on backend/eventloop/shared_resource_loop/sharedresourceloop.go:
```diff
@@ -334,23 +394,585 @@ func processSharedResourceMessage(ctx context.Context, msg sharedResourceLoopMes
+			// If the APICRToDBMapping points to a RepositoryCredential that doesn't exist, delete the APICRToDBMapping
+			// and return an error.
+
+			l.V(sharedutil.LogLevel_Debug).Info("Deleting APICRToDBMapping that points to an invalid repository credential")
```
log.error this

jgwest commented on backend/eventloop/shared_resource_loop/sharedresourceloop.go:
```diff
@@ -334,23 +394,585 @@ func processSharedResourceMessage(ctx context.Context, msg sharedResourceLoopMes
+func createRepoCredOperation(ctx context.Context, dbRepoCred db.RepositoryCredentials, clusterUser *db.ClusterUser, ns string,
+	dbQueries db.DatabaseQueries, apiNamespaceClient client.Client, shouldWait bool, l logr.Logger) (error, string) {
+
+	l.Info("Creating operation...")
```
If you want leave this Info statement in, you should provide more context in the log statement message

jgwest reviewed (changes requested):
A couple general comments in additional to code level comments: 

**Logging**:

As an FYI, my opinionated opinion ( :smiley: ) is that the DEBUG log statements you are using are a bit too verbose, and are missing context that would be beneficial to other developers using them: they are more like debug statements that have been converted. Log statements (at any log level) should generally communicate what action is taking place, why, and on what. 

To pick a random example, this debug statement is missing information about which CR is attempting to be retrieved, and why.
```
l.V(sharedutil.LogLevel_Debug).Info("Attempting to get the CR")
```

Re: when to use which log level, I wrote a tiny bit about this here: https://docs.google.com/document/d/18yjqvgH44KQK7WHSxAAJfmeT2HHO9NAbqD0WFmwSEz8/edit# (search for 'Log statement'). My general philosophy is we should only log at INFO level if we are mutating the state of the service (creating/updating/deleting a database or CR)

It also looks like in many cases comments you've converted code comments wholesale into logs statements (and left both the code comment and log statement in place). I'm not 100% convinced this is desirable, but I'm not requesting they be changed as part of this PR. I'd like to see how they look once running in the product, first.

To pick a random example of this:
```go
// 3) Identify APICRToDBMappings that refer to K8s resources that no longer exist, and find the DTAM that
// refers to the resource we got the request for.
l.V(sharedutil.LogLevel_Debug).Info("Identify APICRToDBMappings that refer to K8s resources that no longer exist, and find the DTAM that refers to the resource we got the request for.")
```

You may also want to look at how we are doing logging in the rest of the codebase, as some of your use of logging seems out of sync with this (for example, logging at debug level, when logging at info or error is what is usually done.)

Likewise, note that `l.WithValues()` doesn't actually log anything, nor does it mutate the `l` object. See other uses in the codebase for examples.


**E2E test requirements**:

So we are on the same page re: what E2E tests need to be done for this story, as we have briefly discussed this in standup previously.

We DO still need the E2E tests as part of this story (it's part of the done criteria for the story). It's true that they won't run in the CI as part of the Story, BUT they are still required to be written for the story (and the PR reviewer should be able to run them locally). With few exceptions, E2E/unit tests should always be delivered with the story before the story is closed.

We don't need the tests to necessarily be of THIS PR, but the lack of E2E tests will still keep the original story (GITOPSRVCE-96) open until the tests are completed.

How should the tests look? I should be able to `GITHUB_USERNAME=jgwest GITHUB_PERSONAL_ACCESS_TOKEN=(a PAT from my username) make test-e2e` (for example), and the private repository credential E2E test should run and pass. 

I've opened a new story, [GITOPSRVCE-303](https://issues.redhat.com/browse/GITOPSRVCE-303), for enabling the test to run in CI (after the E2E test is written), which is NOT required for GITOPSRVCE-96 to close.


I'd expect the E2E test to do this:

(This looks similar/same as what you talk about in `m10-demo`, so it looks like we are on the same page.)

Create these 3 resources (using Go code):

```yaml
apiVersion: managed-gitops.redhat.com/v1alpha1
kind: GitOpsDeployment

metadata:
  name: managed-environment-gitops-depl
  namespace: jane 

spec:

  # Application/component to deploy
  source:
    repoURL: https://github.com/managed-gitops-test-data/private-repo-test
    path: /

  destination: {}

  # Only 'automated' type is currently supported: changes to the GitOps repo immediately take effect (as soon as Argo CD detects them).
  type: automated
```

```yaml
apiVersion: managed-gitops.redhat.com/v1alpha1
kind: GitOpsDeploymentRepositoryCredentials
metadata:
  Name: private-repo-creds
spec:
  url: https://github.com/managed-gitops-test-data
  secret: private-repo-creds-secret
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: private-repo-creds-secret
stringData:
  url: https://github.com/managed-gitops-test-data
  username: ( READ FROM GITHUB_USERNAME ENV VAR )
  password: ( READ FROM GITHUB_PERSONAL_ACCESS_TOKEN ENV VAR)
```

Then, once those 3 are created, wait for the `ConfigMap` named [https://github.com/managed-gitops-test-data/private-repo-test/blob/main/configmap.yaml](config-map-in-private-repo) to be deployed.

drpaneas commented on backend/eventloop/shared_resource_loop/sharedresourceloop.go:
```diff
@@ -334,23 +394,585 @@ func processSharedResourceMessage(ctx context.Context, msg sharedResourceLoopMes
+				}
+
+			} else {
+				l.V(sharedutil.LogLevel_Debug).Info("[Leftover] RepositoryCredential row found in DB")
```
`[Leftover] RepositoryCredential row found in DB` means there is no equivalent CR for this. I will remove this since it's not clear.

drpaneas commented on backend/eventloop/shared_resource_loop/sharedresourceloop.go:
```diff
@@ -334,23 +394,585 @@ func processSharedResourceMessage(ctx context.Context, msg sharedResourceLoopMes
+		// Secret exists, so get its data
+		l.V(sharedutil.LogLevel_Debug).Info("Secret exists, so get its data")
+
+		// TODO: Decide if we need to privateURL = string(secret.Data["url"]) or not
```
See GITOPSRVCE-305

drpaneas commented on backend-shared/config/db/gitopsengineinstance.go:
```diff
@@ -205,7 +206,8 @@ func (dbq *PostgreSQLDatabaseQueries) internalDeleteGitopsEngineInstanceById(ctx
 	if err != nil {
-		return 0, fmt.Errorf("error on deleting operation: %v", err)
+		pgErr := err.(pg.Error)
+		return 0, fmt.Errorf("error on deleting operation: %v\nPGError:%v\n", err, pgErr)
```
if my memory serves me right, I think the `err` is very generic, while the `pg.Error` gives you a the real problem. e.g. that's how I found out about the SSH key varchar issue. See: https://pkg.go.dev/github.com/go-pg/pg#Error

and their example as well, per the documentation:

```go

flight := &Flight{
	Id: 123,
}
err := pgdb.Insert(flight)
panicIf(err)

err = pgdb.Insert(flight)
if err != nil {
	pgErr, ok := err.(pg.Error)
	if ok && pgErr.IntegrityViolation() {
		fmt.Println("flight already exists:", err)
	} else {
		panic(err)
	}
}

// Output:

// flight already exists: ERROR #23505 duplicate key value violates unique constraint "flights_pkey"
````

drpaneas commented on backend/eventloop/shared_resource_loop/sharedresourceloop.go:
```diff
@@ -334,23 +394,585 @@ func processSharedResourceMessage(ctx context.Context, msg sharedResourceLoopMes
+	}
+}
+
+func cleanRepoCredOperation(ctx context.Context, dbRepoCred db.RepositoryCredentials, clusterUser *db.ClusterUser, ns string,
```
Actually I had a test there and I removed it due to the bug. This is how I found the bug `operations.CleanupOperation` bug in the first place. I have now added the test back and i have commented out the line that triggers the bug. I am using a tmp workaround until it's fixed.

drpaneas commented:
Hey @jgwest thanks once again for your time reviewing this. 

### Logging
I agree with all of your comments related to logs. For some reason I had fixed all of those but it must had been a problem with rebasing (I cannot remember how many times I have rebased that PR) that brought some of those back, left there some others and created a general confusion. Thanks for spotting those.

I hope I've addressed all of those comments by now related to the logs.

If you still find any logging issues, by all means, please feel free to commit your desired changes directly. Leaving too many individual review comments makes GitHub UI problematic as it auto-hides them randomly per page refresh, scrolling, etc (so it's easy to miss some).

---

### E2E testing

Sorry for the misunderstanding here. From the discussion we had, I was under the (wrong) impression that providing detailed instructions including commands and screenshots, is actually providing an E2E test, so the developer can run them  manually in their laptop. For example if someone wants to try this locally, they can already just`kubectl apply -f demo{1,2}.yaml` using their GitHub passwords. What is not ready is the automation of that. I was under the impression we will handle this in a different story. My bad :(

Let's handle this in another PR though, because this one is getting complicated.

drpaneas commented:
@jgwest I've added the E2E tests to be part of the PR. Please review and let me know if everything looks fine this time :)

drpaneas commented:
/test managed-gitops-e2e-tests

jgwest reviewed (approved):
LGTM, and great to get this in, this is one of the core features of the GitOps Service that is a must have.  :tada: :tada: :tada:
